### PR TITLE
Add Tron resource staking

### PR DESCRIPTION
**Pull request by peachbits**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203880380954144

### CODE REVIEW

File: src/tron/tronEngine.ts [hunk 1]
```diff
@@ -896,6 +911,93 @@ export class TronEngine extends CurrencyEngine<TronTools> {
...
+    const {
+      params: { nativeAmount, resource }
+    } = action
+
+    const contractJson = {
+      parameter: {
+        value: {
+          owner_address: base58ToHexAddress(this.walletLocalData.publicKey),
+          frozen_balance: parseInt(nativeAmount),
+          frozen_duration: 3,
```
Comment (swansontec) [inline src/tron/tronEngine.ts:993]: Where does this number come from?

File: src/tron/tronTypes.ts [hunk 2]
```diff
@@ -73,6 +73,34 @@ export interface TronTransferParams {
   note?: string
 }
 
+const asResource = asValue('BANDWIDTH', 'ENERGY')
+type Resource = ReturnType<typeof asResource>
+
+export interface TronFreezeAction {
+  name: 'add'
```
Comment (swansontec) [inline src/tron/tronTypes.ts:80]: Please call this `type` instead of `name`.

File: src/tron/tronTypes.ts [hunk 3]
```diff
@@ -73,6 +73,34 @@ export interface TronTransferParams {
...
+export const asTronFreezeAction = asObject<TronFreezeAction>({
+  name: asValue('add'),
+  params: asObject({
+    nativeAmount: asString,
+    resource: asResource
+  })
+})
+
+export interface TronUnfreezeAction {
+  name: 'remove'
```
Comment (swansontec) [inline src/tron/tronTypes.ts:93]: Also change this one to `type: 'remove'`

File: src/tron/tronEngine.ts [hunk 4]
```diff
@@ -527,6 +530,76 @@ export class TronEngine extends CurrencyEngine<TronTools> {
...
+
+        if (
+          hexToBase58Address(fromAddress) !== this.walletLocalData.publicKey
+        ) {
+          break
+        }
+
+        const feeNativeAmount = retArray[0].fee.toString()
+        const nativeAmount = add(frozenAmount.toString(), feeNativeAmount)
+        const currencyCode = this.currencyInfo.currencyCode
```
Comment (swansontec) [inline src/tron/tronEngine.ts:551]: `const {currencyCode} = this.currencyInfo`

**Review by swansontec — CHANGES_REQUESTED** [review]
This looks pretty reasonable. I only see minor tweaks needed.

File: src/tron/tronEngine.ts [hunk 5]
```diff
@@ -246,18 +246,18 @@ export class TronEngine extends CurrencyEngine<TronTools> {
...
       )
-      const balances = asMaybe(asTRXBalance)(res)
-
-      if (balances != null) {
-        this.updateBalance(
-          this.currencyInfo.currencyCode,
-          balances.balance.toString()
-        )
-      } else if (typeof res === 'object' && Object.keys(res).length === 0) {
+      if (typeof res === 'object' && Object.keys(res).length === 0) {
```
[the remark this answers is not in the document]
Comment (peachbits) [reply src/tron/tronEngine.ts:271]: It's an empty object {} so I'll use an asMaybe cleaner against the successful response and check == null

File: src/tron/tronEngine.ts (hunk 1 above)
Comment (peachbits) [reply src/tron/tronEngine.ts:993]: The the network enforced minimum duration in days. I'll add to networkInfo.

File: src/tron/tronTypes.ts (hunk 2 above)
Comment (peachbits) [reply src/tron/tronTypes.ts:80]: Ok will update

File: src/tron/tronEngine.ts (hunk 4 above)
Comment (peachbits) [reply src/tron/tronEngine.ts:551]: I found a few more of these so they're all fixed in the fixup

**Review by swansontec — APPROVED** [review]

**Pull request merged**